### PR TITLE
remove typet::subtype()

### PR DESCRIPTION
**Pull request by kroening**
This adds specialised variants of `typet::subtype()` to the three inheriting classes that use the method.  These cannot use `type_with_subtypet` since C++ doesn't have mixins.  The plan is to remove these methods in favor of the existing type-specific accessor methods.

After that, `typet::subtype()` can be removed.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

File: src/util/c_types.h
Comment (thomasspriggs) [inline src/util/c_types.h:59]: This function should include a call to `bitvector_typet::check` as well, because the new function hides the existing one.

**Review by thomasspriggs — COMMENTED** [review]
I think the missing call to `bitvector_typet::check` should ideally be fixed before this is merged.

You mention that a lack of mixins prevents usage of `type_with_subtypet`, which presumably results in the need to write duplicate definitions of the `subtype()` member functions. I think the Curiously Recurring Template Pattern (CRTP) could be used add in `subtype()` member functions to the classes in the hierarchy which still require them instead. However that would be a different refactor which I do not think should hold up this PR.

**Review by thomasspriggs — APPROVED** [review]

**Pull request merged**
